### PR TITLE
Extract json individual array elements from json index for the transform function jsonExtractIndex

### PR DESCRIPTION
**Pull request by wirybeaver**
Take an example, the json records is shown as follows:
```
    String[] records = {
        "{\"foo\":[{\"bar\":[\"x\",\"y\"]},{\"bar\":[\"a\",\"b\"]}],\"foo2\":[\"u\"]}",
        "{\"foo\":[{\"bar\":[\"y\",\"z\"]}],\"foo2\":[\"u\"]}"
    };
```

Given a json path `$.foo[0].bar[1]`, we want to get extract all values under this json path and associated docId (not the flattened docID) as follows: `{"y":[0], "z":[1]}`.

When setDisableCrossArrayUnnest(true), the conceptual flattened doc model is listed below
Flatten Row  | Flatten ID | Doc Id
-- | -- | --
{ foo[0].bar[0]:x, foo2[0]:u} | 0 | 0
{ foo[0]bar[1]:y, foo2[0]:u} | 1 | 0
{foo[1]bar[0]:a, foo2[0]:u} | 2 | 0
{foo[1]bar[0]:b, foo2[0]:u} | 3 | 0
{ foo[0].bar[0]:y, foo2[0]:u} | 4 | 1
{ foo[0].bar[1]:z, foo2[0]:u} | 5 | 1

The data structure of json indexing below
<a href="https://imgur.com/FEzpKz8"></a>

**Solution**
Step 1: Get the flattenDocIds associated to the given json path. Basically, we will reuse the existing code to process json path containing array index. Use the example above, the array index keys will be `.foo.$index</u0000>0` and `.foo..bar.$index</u0000>1`, bitmap = the conjunction of {0,1,4,5} and {1,3,5} = {1,5}

Step 2: Traverse the inverted indexing dictionary, find all keys with the prefix `.foo..bar.</u0000>`. If the associated bitmap intersecting with the bitmap coming from step 1 is not empty, then parse the value from the key and add the intersection result into the result map.
```
.foo..bar.</u0000>a -> {2}; intersect {1,5} => empty
.foo..bar.</u0000>b -> {3}; intersect {1,5} => empty
.foo..bar.</u0000>x -> {0}; intersect {1,5} => empty
.foo..bar.</u0000>y -> {1,4}; intersect {1,5} => {1}; 
.foo..bar.</u0000>z -> {5}; intersect {1,5} => {5};
```

Step 3: transfer the flattened docId into the original docId by looking up the dodIdMapping
{"y":1, "z": 5} => {"y":0, "z": 1}

### CODE REVIEW

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/json/MutableJsonIndexImpl.java [hunk 1]
```diff
@@ -209,41 +210,11 @@ private RoaringBitmap getMatchingFlattenedDocIds(Predicate predicate) {
     } else {
       key = JsonUtils.KEY_SEPARATOR + key;
     }
-
```
Comment (wirybeaver) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/json/MutableJsonIndexImpl.java:212]: Move the processing of array index to the static function processArrayIndex for code reusing

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/json/ImmutableJsonIndexReader.java [hunk 2]
```diff
@@ -160,94 +161,26 @@ private MutableRoaringBitmap getMatchingFlattenedDocIds(Predicate predicate) {
...
-      // - JSONPath format (e.g. "$.a[1].b"='abc', "$[0]"=1, "$"='abc')
-      // - Legacy format (e.g. "a[1].b"='abc')
-      if (key.charAt(0) == '$') {
+      if (key.startsWith("$")) {
         key = key.substring(1);
       } else {
         key = JsonUtils.KEY_SEPARATOR + key;
       }
-
-      // Process the array index within the key if exists
```
Comment (wirybeaver) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/json/ImmutableJsonIndexReader.java:175]: Keep the key preprocessing but move the generation of matchingDocIds over array index json path to the static function processArrayIndex.

**Comment by Jackie-Jiang** [comment]
@itschrispeck Can you please help take a look?

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/json/ImmutableJsonIndexReader.java [hunk 3]
```diff
@@ -309,12 +242,21 @@ private int getDocId(int flattenedDocId) {
...
+      return matchingDocsMap;
+    }
     int[] dictIds = getDictIdRangeForKey(key);
-
     for (int dictId = dictIds[0]; dictId < dictIds[1]; dictId++) {
       // get docIds from posting list, convert these to the actual docIds
       ImmutableRoaringBitmap flattenedDocIds = _invertedIndex.getDocIds(dictId);
-      PeekableIntIterator it = flattenedDocIds.getIntIterator();
+      PeekableIntIterator it = arrayIndexFlattenDocIds == null ? flattenedDocIds.getIntIterator()
+              : intersect(arrayIndexFlattenDocIds.clone(), flattenedDocIds);
```
Comment (itschrispeck) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/json/ImmutableJsonIndexReader.java:256]: `RoaringBitmap.and(arrayIndexFlattenDocIds, flattenedDocIds).getIntIterator()`

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/json/MutableJsonIndexImpl.java [hunk 4]
```diff
@@ -342,6 +323,51 @@ public String[] getValuesForKeyAndDocs(int[] docIds, Map<String, RoaringBitmap>
     return values;
   }
 
+  private Pair<String, RoaringBitmap> processArrayIndex(String key) {
```
Comment (itschrispeck) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/json/MutableJsonIndexImpl.java:326]: `processArrayIndex` -> `getKeyAndFlattenedDocIds`?

Can you add a comment here about input/output (already done in the immutable reader)

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/json/MutableJsonIndexImpl.java [hunk 5]
```diff
@@ -301,12 +272,22 @@ public Map<String, RoaringBitmap> getMatchingDocsMap(String key) {
...
+      }
       for (Map.Entry<String, RoaringBitmap> entry : _postingListMap.entrySet()) {
         if (!entry.getKey().startsWith(key + BaseJsonIndexCreator.KEY_VALUE_SEPARATOR)) {
           continue;
         }
-        MutableRoaringBitmap flattenedDocIds = entry.getValue().toMutableRoaringBitmap();
-        PeekableIntIterator it = flattenedDocIds.getIntIterator();
+        RoaringBitmap kvPairFlattenedDocIds = entry.getValue();
+        PeekableIntIterator it = arrayIndexFlattenDocIds == null ? kvPairFlattenedDocIds.getIntIterator()
+                : intersect(arrayIndexFlattenDocIds.clone(), kvPairFlattenedDocIds).getIntIterator();
```
Comment (itschrispeck) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/impl/json/MutableJsonIndexImpl.java:287]: `RoaringBitmap.and(arrayIndexFlattenDocIds, flattenedDocIds).getIntIterator()`

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/json/ImmutableJsonIndexReader.java [hunk 6]
```diff
@@ -370,6 +312,90 @@ private int[] getDictIdRangeForKey(String key) {
     return new int[]{minDictId, maxDictId};
   }
 
+  // If key doesn't contain the array index, return <original key, null bitmap>
+  // Elif the key, i.e. the json path provided by user doesn't match any data, return <null, empty bitmap>
+  // Else, return the json path that is generated by replacing array index with . on the original key
+  // and the associated flattenDocId bitmap
```
Comment (itschrispeck) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/json/ImmutableJsonIndexReader.java:318]: nit: javadoc formatting

File: pinot-core/src/test/java/org/apache/pinot/core/operator/transform/function/BaseTransformFunctionTest.java [hunk 7]
```diff
@@ -158,6 +159,11 @@ public void setUp()
               + "\"stringVal\":\"%s\"}", RANDOM.nextInt(), RANDOM.nextLong(), RANDOM.nextFloat(), RANDOM.nextDouble(),
           BigDecimal.valueOf(RANDOM.nextDouble()).multiply(BigDecimal.valueOf(RANDOM.nextInt())),
           df.format(RANDOM.nextInt() * RANDOM.nextDouble()));
+      _jsonArrayValues[i] = String.format(
+          "{\"intVals\":[%s,%s], \"longVals\":[%s,%s], \"floatVals\":[%s,%s], \"doubleVals\":[%s,%s], "
+          + "\"bigDecimalVals\":[%s,%s], \"stringVals\":[\"%s\",\"%s\"]}",
+          0, 1, 0L, 1L, 0.0f, 1.0f, 0.0, 1.0, BigDecimal.valueOf(0.0), BigDecimal.valueOf(1.0),
+          df.format(0.0), df.format(1.0));
```
[the remark this answers is not in the document]
Comment (wirybeaver) [reply pinot-core/src/test/java/org/apache/pinot/core/operator/transform/function/BaseTransformFunctionTest.java:166]: will do

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/json/ImmutableJsonIndexReader.java (hunk 3 above)
Comment (wirybeaver) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/json/ImmutableJsonIndexReader.java:256]: RoaringBitmap.and require the input type is RoaringBitmap();
MutableRoaringBitmap and ImmutableRoaringBitamp are not the subclass of RoaringBitmap.

**Comment by wirybeaver** [comment]
The integration test fails on a mutlistage query irrelevant to json. the temeurin-11 succeeds.

**Comment by Jackie-Jiang** [comment]
@saurabhd336 Is this PR trying to solve the same problem as yours (#12532)?

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/json/ImmutableJsonIndexReader.java (hunk 3 above)
Comment (itschrispeck) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/json/ImmutableJsonIndexReader.java:334]: How about `MutableRoaringBitmap.and(arrayIndexFlattenDocIds, flattenedDocIds)`?

**Review by itschrispeck — APPROVED** [review]

**Review by Jackie-Jiang — APPROVED** [review]

**Pull request merged**